File: advent of code/2022/deno/day6.ts
import { readInput } from "./utils.ts"

export function day6(test = false) {
  const input = readInput(6, test).split("")

  const part1 = input.findIndex((_, i) => input.slice(i, i + 4).isUnique()) + 4
  const part2 = input.findIndex((_, i) => input.slice(i, i + 14).isUnique()) + 14
  return [part1, part2]
}

File: advent of code/2022/ocaml/lib/day6.ml
open Core

let is_uniq (s : string) =
  let str_len = String.length s in
  let set_len = String.to_list s |> Char.Set.of_list |> Set.length in
  str_len = set_len

let find_uniq_slice_exn (s : string) (len : int) =
  String.findi s ~f:(fun i _ -> String.slice s i (i + len) |> is_uniq)
  |> Option.value_exn |> Tuple2.get1

let solve ~(file : string) =
  let input = Utils.read_file file in
  let part1 = 4 + find_uniq_slice_exn input 4 in
  let part2 = 14 + find_uniq_slice_exn input 14 in
  (part1, part2)

let%test_unit _ = [%test_eq: int * int] (solve ~file:"day6-example.txt") (7, 19)
